docs: network changes (#4)
opengovsg/im8-cfa-sandbox: networking.md

# Networking

### Network connectivity

The system **must not** be connected to the Government Enterprise Network (GEN), or to other CII and SII systems

> Internet access **must** be blocked by default, with granular whitelisting for individual services that require Internet connectivity (e.g. through a network firewall with URL or IP whitelist)

Reviewer: Is this guideline for incoming traffic only? My understanding is that an app server deployed to a private subnet with a NAT Gateway is able to initiate connections to any external endpoint.

Reviewer 2: This is more of referring to accessing the Internet through outgoing Internet traffic. Maybe we can sharpen this articulation. By blocking outgoing traffic by default, it can mitigate against exploitation of two stage vulnerabilities like Log4J.

Reviewer: I am wondering how to achieve this in a way that is not overly onerous to the system. The default way is probably to modify the assigned security group of the app server to include the outgoing connection's target IPs, but this would obstruct API integration with external SaaS providers that do not commit to providing static IPs for integration against.

Author: My mistake on the unclear wording. What do we think of this change?

```suggestion
Outgoing Internet access **must** be blocked by default, with granular whitelisting for individual services that require Internet connectivity (e.g. through a network firewall with URL or IP whitelist)
```

Reviewer: May I check what are the concrete steps that one may take to achieve domain whitelisting?

I am worried that this is non-trivial to achieve; and end up impose significant overheads on what could have been relatively straightforward SaaS integration.

Author: It used to be sticking a forward proxy (running squid or similar in a VM) was the only option in policy.

It doesn't have to follow as we develop new policy. It could be as basic as a default deny Security Group with destination IPs whitelisted.

For more advanced cases, an AWS network firewall running like a forward proxy can be utilised to cover more granular URL based filtering (as well as network inspection). It is more complex but not too difficult to automate (unlike Squid based solution). At scale, it also has the benefit of becoming a shared service for all accounts in an org to use as central egress and NAT (on the roadmap for GCC)

Reviewer 2: Thanks @liangyuanruo and @hunter.

This requirement is to solve insecure AWS default where the tradeoff inclines towards operation convenience. Yes, there is limitation on Security Groups and NACL w.r.t. domains whitelisting. Agree that this may impose some issues when applications integrate with SaaS that does not have fixed IP address range. 

On this, CSG with inputs from CODEX, GDS, and AWS colleagues have co-developed an IAC that has incorporated implementation of AWS network firewall to achieve this. This template is currently hosted on SHIP-HATS which I can share for more inputs. This template is currently scheduled for VAPT as we also intend to share publicly with WOG to help expedite Agencies' deployment.

Alternatively, if we were to take some risk, the whitelisting can be performed at IP address ranges that the SaaS uses. Then good to have this risk documented as part of standard process for informed decision.

Reviewer: I see. Is it a reasonable expectation for a dev team to have to find out all the domain and IP addresses that a SaaS provider uses though, especially if the information may not have been published?

As a concrete example, FormSG uses Twilio for SMSes and DataDog for logging. These services provide an SDK for easy integration, but (afaik) their documentation do not provide any information about the specific domains or IPs that the SDK will be calling. Even if it were figured out, this may also be subject to unannounced change as the SDKs get upgraded or if the SaaS providers make infrastructural changes.

The only exceptions I can tell are for webhook callbacks, where an authentication scheme is usually communicated (signed headers, maybe fixed IPs), but these are connections initiated externally.

Reviewer: Thanks @ronghwa80 for weighing in - seems that it may be premature to mandate this? Perhaps we should reduce the requirement to "recommend" instead?

Author: I'd be leaning towards a mandate for limiting egress since it can mitigate a significant number of risks if an application is compromised. I think in the scope of the sandbox with using only managed services then it might be possible to recommend until we find a usable.

> Is it a reasonable expectation for a dev team to have to find out all the domain and IP addresses that a SaaS provider uses though, especially if the information may not have been published?

I do think it is reason to expect developers to know how their app is communicating. Even without considering security, knowing where an App needs to make external network requests can help with debugging and reliable operations (eg. Twilio is message delivery is slow and no reported outage, lets test the endpoint)

On application architecture specifics...

IPs are not a great means of egress access control since they can change and these days not a contiguous range for simple management but it is pretty easy to implement. Some services do publish (or discoverable from network logs) and it might be useful for the basics.

Glad to see that the IaC templates are including Network Firewall which should help simplify implementation. @ronghwa80 I'm looking forward to the team [innersourcing](https://www.developer.tech.gov.sg/guidelines/standards-and-best-practices/innersource.html) it 😉

Route53 DNS resolver firewalling is another alternative to add another layer of defence. Somewhat in-between a IP and a full URL firewall.

As something to sidestep the direct network egress (and related to the [comment](https://github.com/opengovsg/im8-cfa-sandbox/pull/4#issuecomment-1131806446) on CSP services), using an indirect method might be an alternative. eg. SNS topics triggering a Lambda to post to Twilio, streaming CW logs to Datadog via Kinesis, etc.

A few things to consider to add more flexibility (but also not something we want to encode in the guide... otherwise we end up like AIAS ;)

> Outgoing Internet access **must** be blocked by default, with granular whitelisting for individual services that require Internet connectivity (e.g. through a network firewall with URL or IP whitelist)

Author: Change to 'should' based on discussions

```suggestion
Outgoing Internet access should be blocked by default, with granular whitelisting for individual services that require Internet connectivity (e.g. through a network firewall with URL or IP whitelist)
```

### Network environments

> Segregation of cloud resources **must** be achieved between environments (staging, UAT, production, etc), via separate accounts.

Reviewer: I am wondering if it is strictly necessary to _mandate_ complete segregation of cloud accounts for environments, over an alternative such as having separate VPCs? I'll admit that there are probably some security benefits from this as a default practice - such as simplifying the writing of IAM policies. If so perhaps we could expand this to include the justification.

Reviewer 2: https://docs.aws.amazon.com/whitepapers/latest/organizing-your-aws-environment/benefits-of-using-multiple-aws-accounts.html

By applying different account for different environment helps to limit the scope of impact from adverse events. This aligns with the assume breach mindset. :)

An AWS account provides security, access, and billing boundaries for your AWS resources that can help you achieve resource independence and isolation. This isolation boundary provides you with a way to limit the risks of an application-related issue, misconfiguration, or malicious actions. If an issue occurs within one account, impacts to workloads contained in other accounts can be either reduced or eliminated.

Author: My initial edit of that line had a split between prod and non-prod. Production is usually the most critical for isolation with the remaining environments less so and could be combined. I felt the wording may end up with too many questions so opted to keep it simple but with a wider scope.

Although not a obvious security benefit, using account based environments means that quotas are not combined in the single account case, which could lead to denial of service (AWS API limits hit, unable to autoscale a resource, etc). Operationally, it also can have benefits for applying different AWS support tiers to different accounts.

### Network segmentation

Resources **must** be deployed into an appropriate types of network and identity based segments (e.g. subnet, VPC, IAM role, etc). As a minimum, the following segments should be considered:

* **Public** with an attached Internet Gateway suitable for external-facing resources e.g. load balancers for HTTP traffic
* **Private** without an Internet Gateway, but have resources that may initiate access to the internet via a NAT Gateway e.g. application servers
* **Data** without an Internet Gateway and NAT Gateway, that are not suitable for direct external access (e.g. persistence layers such as databases, caches, etc.)

> Individual resources in a segment should be further isolated from others by identity and network based controls (e.g. IAM roles, Security Groups, etc) to minimise pivot risk. 

Reviewer: May I ask what does "pivot risk" mean?

Author: In this case, an app server could be compromised and an attack launched from that server to other servers in the autoscaling group (or anything else accessible in the segment).

Reviewer: I see, I am unsure as to whether this is a common term but I don't want to be pedantic 😅

Author: It is a common term. Alternatively, "Lateral movement" might be another option (and perhaps better as less specific to a technique). @ronghwa80, any preference?

Reviewer 2: @hunter, yes, I was thinking that you might be referring to lateral movement, which is a more generic term.

Agree with you that identity-based security in conjunction with network segmentation would form the fundamental of GTbA to limit the blast radius significantly. That said, I want to highlight that I agree with your approach of listing IAM roles as an example. It is good to provide some flexibility on how identity can be implemented, e.g. for certain implementation, the use of secret management tools like AWS secret manager will work as effectively.

Author: ```suggestion
Individual resources in a segment should be further isolated from others by identity and network based controls (e.g. IAM roles, Security Groups, etc) to limit lateral movement. 
```

### Network peering

> Where available, VPC private endpoints should be used for cross-service communication (including 3rd-party service providers) to avoid having traffic crossing the open internet. For cases that don't support private endpoints (eg. bi-directional traffic, legacy services, etc), VPC Peering or Transit Gateways (TGW) should be adopted.

Reviewer: Good suggestion - I was not aware of VPC PrivateLink on AWS, but it appears to be a superior solution compared to VPC Peering because it constrains traffic to allowed endpoints.
